lib/charms/rolling_ops/v0/rollingops.py

```diff
@@ -0,0 +1,353 @@
+            self.relation.data[self.unit].update({"status": status})
+
+        if status == self.RELEASE:
+            self.relation.data[self.unit].update({"status": status})
+
+        if status == self.GRANTED:
+            self.relation.data[self.app].update({self.unit: status})
+
+        if status is self.IDLE:
+            self.relation.data[self.app].update({self.unit: status})
+
+    def acquire(self):
+        self.status = self.ACQUIRE
+
+    def release(self):
+        self.status = self.RELEASE
+
+    def clear(self):
+        self.status = self.IDLE
+
+    def grant(self):
+        self.status = self.GRANTED
+
+    def is_held(self):
+        return self.status == self.GRANTED
+
+    def release_requested(self):
+        return self.status == self.RELEASE
+
+    def is_pending(self):
+        return self.status == self.ACQUIRE
+
+
+class Locks:
+    """Generator that returns a list of locks."""
+    def __init__(self, rel_name, charm):
+        self.rel_name = rel_name
+        self.charm = charm
+
+        # Gather all the units.
+        relation = charm.model.relations[rel_name][0]
+        units = [unit for unit in relation.units]
+
+        # Plus our unit ...
+        units.append(charm.framework.model.unit)
+
+        self.units = units
+
+    def __iter__(self):
+
+        for unit in self.units:
+            yield Lock(self.rel_name, self.charm, unit=unit)
+
+
+class RollingEvent(EventBase):
+    def __init__(self, handle, name):
+        super().__init__(handle)
+        self._name = name
+
+    def snapshot(self):
+        return {"name": self._name}
+
+    def restore(self, snapshot):
+        self._name = snapshot["name"]
+
+    @property
+    def name(self):
+        """The `name` property allows us to namespace Rolling Operations."""
+        return self._name
+
+
+class RunWithLock(RollingEvent):
+    pass
+
+
+class AcquireLock(RollingEvent):
+    pass
+
+
+class ProcessLocks(RollingEvent):
+    pass
+
+
+class RollingEvents(CharmEvents):
+    run_with_lock = EventSource(RunWithLock)
+    acquire_lock = EventSource(AcquireLock)
+    process_locks = EventSource(ProcessLocks)
+
+
+class RollingOpsManager(Object):
+    """Emitters and handlers for rolling ops."""
+
+    def __init__(self, charm, rel_name, func_run, event_acquire):
+        """
+        params:
+            charm: the charm we are attaching this to.
+            rel_name: an identifier, by convention based on the name of the relation in the
+                metadata.yaml, which identifies this instance of RollingOperatorsFactory,
+                distinct from other instances that may be hanlding other events.
+            func_run: a closure to run when we have a lock. (It must take a CharmBase object and
+                EventBase object as args.)
+            event_acquire: The event that will prompt us to acquire a lock.
+        """
+        super().__init__(charm, None)
+
+        self.charm = charm
+        self.name = rel_name
```

Comment:
Might it be worth doing a rudimentary check against the charm's metadata to ensure that the relation name specified exists, and is a peer relation?